dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,519 @@
+        query = []
+        if parent is not None:
+            query.append("'{}' in parents".format(parent))
+        if name is not None:
+            query.append("name = '{}'".format(name))
+        params = {"q": " and ".join(query), "spaces": self.space}
+        params.update(add_params)
+        while True:
+            data = self.request("GET", "drive/v3/files", params=params).json()
+            for i in data["files"]:
+                yield i
+            if not data.get("nextPageToken"):
+                break
+            params["pageToken"] = data["nextPageToken"]
+
+    def get_metadata_by_path(self, root, path, fields=[]):
+        parent = self.get_metadata_by_id(root)
+        current_path = ["gdrive://" + parent["id"]]
+        parts = path.split("/")
+        # only specify fields for the last search query
+        kwargs = [{}] * (len(parts) - 1) + [
+            {"add_params": {"fields": "files({})".format(",".join(fields))}}
+            if fields
+            else {}
+        ]
+        for part, kwargs in zip(parts, kwargs):
+            if not self.metadata_isdir(parent):
+                raise GDriveError(
+                    "{} is not a folder".format("/".join(current_path))
+                )
+            current_path.append(part)
+            files = list(self.search(parent["id"], part, **kwargs))
+            if len(files) > 1:
+                raise GDriveError(
+                    "path {} is duplicated".format("/".join(current_path))
+                )
+            elif len(files) == 0:
+                raise GDriveResourceNotFound("/".join(current_path))
+            parent = files[0]
+        return parent
+
+    def metadata_isdir(self, metadata):
+        return metadata["mimeType"] == self.MIME_GOOGLE_APPS_FOLDER
+
+    def get_file_checksum(self, path_info):
+        metadata = self.get_metadata_by_path(
+            path_info.root, path_info.path, params={"fields": "md5Checksum"}
+        )
+        return metadata["md5Checksum"]
+
+    def _list_files(self, folder_id):
+        for i in self.search(parent=folder_id):
+            if self.metadata_isdir(i):
+                for j in self._list_files(i["id"]):
+                    yield i["name"] + "/" + j
+            else:
+                yield i["name"]
+
+    def list_cache_paths(self):
+        try:
+            root = self.get_metadata_by_path(self.root, self.prefix)
+        except GDriveResourceNotFound as e:
+            logger.debug("list_cache_paths: {}".format(e))
+        else:
+            for i in self._list_files(root["id"]):
+                yield self.prefix + "/" + i
+
+    @only_once
+    def mkdir(self, parent, name):
+        data = {
+            "name": name,
+            "mimeType": self.MIME_GOOGLE_APPS_FOLDER,
+            "parents": [parent],
+            "spaces": self.space,
+        }
+        return self.request("POST", "drive/v3/files", json=data).json()
+
+    @only_once
+    def makedirs(self, parent, path):
```

Comment:
base class accepts path_info .. is it a bug?

Reply:
Oh, I didn't noticed that the base class even has this method. Yes, it should be fixed, but this would require some work on proper folder IDs caching (which actually should be done anyway). Working on it.